new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,241 @@
+            continue;
+        }
+        if (null === value) {
+            newSearchParams.delete(key);
+        } else {
+            newSearchParams.set(key, String(value));
+        }
+    }
+    if (null === filePath) {
+        newSearchParams.delete(SEARCH_PARAM_NAME.FILE_PATH);
+    } else if ("string" === typeof filePath) {
+        newSearchParams.set(SEARCH_PARAM_NAME.FILE_PATH, filePath);
+    }
+
+    let searchString = newSearchParams.toString();
+    if (!(/%23|%26/).test(searchString)) {
+        searchString = decodeURIComponent(searchString);
+    }
+
+    return searchString;
+};
+
+/**
+ * Computes updated URL hash parameters based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ * @return The updated hash parameters string.
+ */
+const getUpdatedHashParams = (updates: UrlHashParamUpdatesType) => {
+    const newHashParams = new URLSearchParams(window.location.hash.substring(1));
+    for (const [key, value] of Object.entries(updates)) {
+        if (null === value) {
+            newHashParams.delete(key);
+        } else {
+            newHashParams.set(key, String(value));
+        }
+    }
+
+    return newHashParams.toString();
+};
+
+/**
+ * Updates search parameters in the current window's URL based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the search parameters.
+ * If a key's value is `null`, the key will be removed from the search parameters.
+ */
+const updateWindowSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newUrl = new URL(window.location.href);
+    newUrl.search = getUpdatedSearchParams(updates);
+    window.history.pushState({}, "", newUrl);
+};
+
+/**
+ * Updates hash parameters in the current window's URL based on the provided key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the hash parameters.
+ * If a key's value is `null`, the key will be removed from the hash parameters.
+ */
+const updateWindowHashParams = (updates: UrlHashParamUpdatesType) => {
+    const newHash = getUpdatedHashParams(updates);
+    const currHash = window.location.hash.substring(1);
+    if (newHash !== currHash) {
+        const newUrl = new URL(window.location.href);
+        newUrl.hash = newHash;
+        window.history.pushState({}, "", newUrl);
+        window.dispatchEvent(new HashChangeEvent("hashchange"));
+    }
+};
+
+/**
+ * Retrieves all search parameters from the current window's URL.
+ *
+ * @return An object containing the search parameters.
+ */
+const getAllWindowSearchParams = () => {
+    const urlSearchParams: NullableProperties<UrlSearchParams> = structuredClone(
+        URL_SEARCH_PARAMS_DEFAULT
+    );
+    const searchParams = new URLSearchParams(window.location.search.substring(1));
```

Comment:
I feel like the names of these two should be swapped. Ditto in the next function.